File: .config/dmenu/format_merge_lines.py
import glob

def check_first_line(line):
    temp_line = line.strip()
    return (temp_line == '' or temp_line.startswith('/') or temp_line.startswith('*')) and not temp_line.startswith('#')

def check_for_comment(line):
    return line.strip().startswith('/') or line.strip().startswith("*")

def check_for_content(line):
    return line.strip() != '' and not check_for_comment(line) and not line.strip().startswith('#')

def check_for_paren(line):
    return '(' in line.strip()

def check_for_end_paren(line):
    return line.strip().endswith(')')

def check_for_func_start(line):
    return line.replace('\n', '').startswith('{')

def process_file(filename):
    change_counter = 0
    with open(filename, 'r') as file:
        lines = file.readlines()

    new_lines = []
    i = 0
    while i < len(lines):
        if i < len(lines) - 3 and check_first_line(lines[i]) and check_for_content(lines[i+1]) and check_for_end_paren(lines[i+2]) and check_for_func_start(lines[i+3]):
            new_lines.append(lines[i])
            new_lines.append(lines[i+1].rstrip() + ' ' + lines[i+2].rstrip() + '\n')
            new_lines.append(lines[i+3])
            i += 4
            change_counter += 1
        elif i < len(lines) - 4 and check_first_line(lines[i]) and check_for_content(lines[i+1]) and check_for_paren(lines[i+2]) and check_for_end_paren(lines[i+3]) and check_for_func_start(lines[i+4]):
            new_lines.append(lines[i])
            new_lines.append(lines[i+1].rstrip() + ' ' + lines[i+2].rstrip() + '\n')
            new_lines.append(lines[i+3])
            new_lines.append(lines[i+4])
            i += 5
            change_counter += 1
        elif i < len(lines) - 5 and check_first_line(lines[i]) and check_for_content(lines[i+1]) and check_for_paren(lines[i+2]) and check_for_content(lines[i+3]) and check_for_end_paren(lines[i+4]) and check_for_func_start(lines[i+5]):
            print(f"Found function with three lines of arguments, see: {lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4]}")
            new_lines.append(lines[i])
            i += 1
        else:
            new_lines.append(lines[i])
            i += 1

    with open(filename, 'w') as file:
        file.writelines(new_lines)
    return change_counter

# Replace the path with the directory where your .c files are located
path = './' # Current directory
total_change_count = 0
for filename in glob.glob(path + '*.c'):
    print(f'Processing file: {filename}')
    file_change_count = process_file(filename)
    print(f"Done! Changes done: {file_change_count}")
    total_change_count += file_change_count

print(f"Done processing files! Total changes done: {total_change_count}")

File: bin/my_scripts/utils/wip/format_merge_lines.pl
#!/usr/bin/perl

use strict;
use warnings;
use File::Find;
use File::Slurp;

sub check_first_line {
    my ($line) = @_;
    $line =~ s/^\s+|\s+$//g; # Trim whitespaces
    return ($line eq '' || $line =~ m{^/} || $line =~ m{^\*}) && $line !~ m{^#};
}

sub check_for_comment {
    my ($line) = @_;
    $line =~ s/^\s+|\s+$//g;
    return $line =~ m{^/} || $line =~ m{^\*};
}

sub check_for_content {
    my ($line) = @_;
    $line =~ s/^\s+|\s+$//g;
    return $line ne '' && !check_for_comment($line) && $line !~ m{^#};
}

sub check_for_paren {
    my ($line) = @_;
    $line =~ s/^\s+|\s+$//g;
    return $line =~ m{\(};
}

sub check_for_end_paren {
    my ($line) = @_;
    $line =~ s/^\s+|\s+$//g;
    return $line =~ m{\)$};
}

sub check_for_func_start {
    my ($line) = @_;
    $line =~ s/[\n\r]//g; # Remove newlines
    return $line =~ m{^\{};
}

sub process_file {
    my ($filename) = @_;
    my @lines = read_file($filename);
    my @new_lines;
    my $change_counter = 0;
    my $i = 0;

    while ($i < scalar @lines) {
        if ($i < scalar @lines - 3 && check_first_line($lines[$i]) && check_for_content($lines[$i + 1]) && check_for_end_paren($lines[$i + 2]) && check_for_func_start($lines[$i + 3])) {
            push @new_lines, $lines[$i];
            push @new_lines, $lines[$i + 1];
            $new_lines[-1] =~ s/\s+$//; # Trim right whitespace
            $new_lines[-1] .= ' ' . $lines[$i + 2];
            push @new_lines, $lines[$i + 3];
            $i += 4;
            $change_counter++;
        } else {
            push @new_lines, $lines[$i];
            $i++;
        }
    }

    write_file($filename, @new_lines);
    return $change_counter;
}

sub wanted {
    if (/\.c$/ && -f $_) {
        print "Processing file: $_\n";
        my $file_change_count = process_file($File::Find::name);
        print "Done! Changes done: $file_change_count\n";
    }
}

my $dir_path = './'; # Current directory
find(\&wanted, $dir_path);
